Userland/Libraries/LibCore/Process.cpp

```diff
@@ -57,64 +56,105 @@ struct ArgvList {
-    {
+    posix_spawn_file_actions_t spawn_actions;
+    CHECK(posix_spawn_file_actions_init(&spawn_actions));
+    ScopeGuard cleanup_spawn_actions = [&] {
+        posix_spawn_file_actions_destroy(&spawn_actions);
+    };
+
+    if (options.working_directory.has_value()) {
 #ifdef AK_OS_SERENITY
-        posix_spawn_file_actions_t spawn_actions;
-        posix_spawn_file_actions_init(&spawn_actions);
-        ScopeGuard cleanup_spawn_actions = [&] {
-            posix_spawn_file_actions_destroy(&spawn_actions);
-        };
-        if (!m_working_directory.is_empty())
-            posix_spawn_file_actions_addchdir(&spawn_actions, m_working_directory.characters());
-
-        auto pid = TRY(System::posix_spawn(m_path.view(), &spawn_actions, nullptr, const_cast<char**>(get().data()), System::environment()));
-        if (keep_as_child == Process::KeepAsChild::No)
-            TRY(System::disown(pid));
+        CHECK(posix_spawn_file_actions_addchdir(&spawn_actions, options.working_directory->characters()));
 #else
-        auto pid = TRY(System::posix_spawn(m_path.view(), nullptr, nullptr, const_cast<char**>(get().data()), System::environment()));
-        // FIXME: Support keep_as_child outside Serenity.
-        (void)keep_as_child;
+        TODO();
 #endif
-        return pid;
     }
-};
+
+    for (auto const& file_action : options.file_actions) {
+        TRY(file_action.visit(
+            [&](FileAction::OpenFile const& action) -> ErrorOr<void> {
+                CHECK(posix_spawn_file_actions_addopen(
+                    &spawn_actions,
+                    action.fd,
+                    action.path.characters(),
+                    File::open_mode_to_options(action.mode | Core::File::OpenMode::KeepOnExec),
+                    action.permissions));
+                return {};
+            }));
+    }
+
+#undef CHECK
+
+    ArgvList argv_list(options.path, options.arguments.size());
+    for (auto const& argument : options.arguments)
+        argv_list.append(argument.characters());
+
+    auto pid = TRY(System::posix_spawn(options.path.view(), &spawn_actions, nullptr, const_cast<char**>(argv_list.get().data()), System::environment()));
+
+    return Process { pid };
+}
 
 ErrorOr<pid_t> Process::spawn(StringView path, ReadonlySpan<DeprecatedString> arguments, DeprecatedString working_directory, KeepAsChild keep_as_child)
 {
-    ArgvList argv { path, arguments.size() };
-    for (auto const& arg : arguments)
-        argv.append(arg.characters());
-    argv.set_working_directory(working_directory);
-    return argv.spawn(keep_as_child);
+    auto process = TRY(spawn({
+        .path = path,
+        .arguments = Vector<DeprecatedString> { arguments },
+        .working_directory = working_directory.is_empty() ? Optional<DeprecatedString> {} : Optional<DeprecatedString> { working_directory },
+    }));
+
+    if (keep_as_child == KeepAsChild::No)
+        TRY(process.disown());
```

Comment:
Food for thought: are we actually using disown anywhere in the system? From what I remember it was an experimental thing from way early in development that isn't really a thing on any other unix.

Reply:
`KeepAsChild::No` is a default option for `keep_as_child`, so it is implicitly used in a lot of places. That said, the only place where we pass `KeepAsChild::No` explicitly is in SystemServer when it spawns `/bin/DeviceMapper`.